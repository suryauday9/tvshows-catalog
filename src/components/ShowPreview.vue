<template>
  <div v-if="show">
      <v-card
        class="ma-1"
        height="180"
        width="180"
        @click="viewDetails(show.id)"
      >
        <v-img
          v-if="show.image != null"
          :src="show.image.medium"
          height="180"
          class="white--text align-end"
        >
          <v-card-title class="subtitle-2 black pa-1">
            <div class="d-flex align-center show-name">
              {{ show.name }}
            </div>
            <div class="d-flex align-center show-rating" v-if="show.rating && show.rating.average">
              {{ show.rating.average }}
              <v-icon color="red">mdi-star</v-icon>
            </div>
          </v-card-title>
        </v-img>
        <v-img
          v-else
          :src="require('../assets/tv-show-img.jpg')"
          height="180"
          class="white--text align-end"
        >
          <v-card-title class="subtitle-2 black pa-3">
            {{ show.name }}
          </v-card-title>
        </v-img>
      </v-card>
  </div>
</template>

<script>
export default {
  name: "ShowPreview",
  props: {
    show: {}
  },
  methods: {
    viewDetails(id) {
      this.$router.push({ name: "show", params: { id: id } });
    }
  }
};
</script>
<style scoped>
.show-name {
  width: 100%;
  display: block !important;
}
.show-rating {
  width: 100%;
  display: block !important;
}
</style>
