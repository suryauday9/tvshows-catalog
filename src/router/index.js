import Vue from "vue";
import VueRouter from "vue-router";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: 'ShowsCatalog',
    component: () => import('../views/ShowsCatalog.vue')
  },
  {
    path: '/show',
    name: 'show',
    component: () =>
      import('../views/ShowDetails.vue'),
    props: true,
  },
  {
    path: '/404',
    alias: '*',
    name: 'notFound',
    component: () =>
      import('../views/Page404.vue'),
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes
});

export default router;
