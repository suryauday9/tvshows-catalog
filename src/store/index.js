import Vue from "vue";
import Vuex from "vuex";

import {
getAllShows,
getShowDetails,
getShowThumbnails,
getShowsBySearch
} from "../services/service";

Vue.use(Vuex);

const state = {
  shows: [],
  genres: [],
  showDetails: [],
  showThumbnails: [],
  showSearchbar: false
};

export const sortByRating = (response) => {
  return response.sort((prevValue, nextValue) =>
  prevValue.rating.average < nextValue.rating.average ? 1 : -1);
};

export const getGenres = (response) => {
  return response
  .reduce((result, show) => result.concat(show.genres), [])
  .filter((genre, index, self) => self.indexOf(genre) === index);
};

export const actions = {
  async getShowslist({ commit }) {
    getAllShows().then((response) => {
      commit('setShowslist', sortByRating(response));
      commit('setGenreslist', getGenres(response));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting all shows: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowDetails({ commit }, showId) {
    getShowDetails(showId).then((response) => {
      commit('setShowDetails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show details: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowThumbnails({ commit }, showId) {
    await getShowThumbnails(showId).then((response) => {
      commit('setShowThumbnails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show thumbnails: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowsBySearch({ commit }, searchTerm) {
    await getShowsBySearch(searchTerm).then((response) => {
      let resultSet = response.map(result => (result.show));
      commit('setShowslist', resultSet);
      commit('setGenreslist', getGenres(resultSet));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem in show search: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  getSearchbarStatus ({ commit }) {
    commit('setSearchbarStatus', !this.state.showSearchbar);
  }
};

export const mutations = {
  setShowslist(state, payload) {
    state.shows = payload;
  },
  setGenreslist(state, payload) {
      state.genres = payload;
  },
  setShowDetails(state, payload) {
    state.showDetails = payload;
  },
  setShowThumbnails(state, payload) {
    state.showThumbnails = payload;
  },
  setSearchbarStatus(state, payload) {
    state.showSearchbar = payload;
  }
};

export default new Vuex.Store({
  state: state,
  actions: actions,
  mutations: mutations,
  modules: {}
});
