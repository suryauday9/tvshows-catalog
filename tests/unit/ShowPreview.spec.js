import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';

import { initializeApp } from './initializeApp';
import { showsData } from './response';
import ShowPreview from '../../src/components/ShowPreview.vue';
import { routes } from '../../src/router';

const vuetify = new Vuetify();
const localVue = initializeApp(createLocalVue());

describe('ShowPreview component', () => {  
    let wrapper;
    const router = new VueRouter(routes);

    const store = new Vuex.Store({
        state: {
            shows: showsData,
            genres: [],
            showDetails: [],
            showThumbnails: [],
            showSearchbar: false
        },
        actions: {}
    });

    const mountComponent = (options) => {
        return shallowMount(ShowPreview, {
            localVue,
            router,
            store,
            propsData: { 'show' : showsData[0] },
            vuetify,
            ...options,
        })
    }

    beforeEach(() => {
        wrapper = mountComponent();
    })

    afterEach(() => {
        wrapper.destroy();
    })

    it('renders component correctly', () => {
        // see if component renders
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should be a Vue instance', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    });

    it('should display show name', () => {
        expect(wrapper.find("div.show-name").text()).toEqual(showsData[0].name);
    });

    it('it should navigate to ShowDetails page', () => {
        wrapper.vm.viewDetails(2);
        expect(wrapper.vm.$router.history.pending.name).toBe('show');
    });

});