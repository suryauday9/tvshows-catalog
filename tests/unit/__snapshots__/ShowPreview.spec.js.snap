// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowPreview component renders component correctly 1`] = `
<div>
  <v-card-stub
    class="ma-1"
    height="180"
    loaderheight="4"
    tag="div"
    width="180"
  >
    <v-img-stub
      class="white--text align-end"
      height="180"
      options="[object Object]"
      position="center center"
      src="https://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg"
      transition="fade-transition"
    >
      <v-card-title-stub
        class="subtitle-2 black pa-1"
      >
        <div
          class="d-flex align-center show-name"
        >
          
            Under the Dome
          
        </div>
         
        <div
          class="d-flex align-center show-rating"
        >
          
            6.5
            
          <v-icon-stub
            color="red"
          >
            mdi-star
          </v-icon-stub>
        </div>
      </v-card-title-stub>
    </v-img-stub>
  </v-card-stub>
</div>
`;
